File: backend/app/api/trace.py
"""数据溯源接口：账→单 / 单→账 双向可查（对应工作包 7 溯源机制）。"""
from __future__ import annotations

from datetime import date, datetime, time

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.database import get_db
from app.models import Bank, BankAccount, BankRawFlow, FlowBatch, FlowReview, FlowValidation, PushRecord, TransFlow

router = APIRouter(prefix="/api/trace", tags=["trace"])


@router.get("/by-voucher")
def by_voucher(
    voucher_no: str = Query(...),
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """账 → 单：金蝶凭证号回溯原始流水、采集批次、复核与推送记录。"""
    push = db.execute(
        select(PushRecord).where(PushRecord.voucher_no == voucher_no).order_by(PushRecord.push_id.desc()).limit(1)
    ).scalar_one_or_none()
    if push is None:
        raise HTTPException(status_code=404, detail="未找到该凭证号对应的流水")
    return _record_trace(db, push.record_id)


@router.get("/by-record")
def by_record(
    record_id: int = Query(...),
    db: Session = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """单 → 账：原始流水回溯批次 / 校验 / 复核 / 推送 / 凭证号。"""
    return _record_trace(db, record_id)


def _record_trace(db: Session, record_id: int) -> dict:
    flow = db.get(TransFlow, record_id)
    if flow is None:
        raise HTTPException(status_code=404, detail="流水不存在")

    batch = db.get(FlowBatch, flow.batch_id)
    bank = db.get(Bank, flow.bank_id)
    account = db.get(BankAccount, flow.account_id)
    raw = db.get(BankRawFlow, flow.raw_id) if flow.raw_id else None
    validations = db.execute(
        select(FlowValidation).where(FlowValidation.record_id == record_id)
    ).scalars().all()
    reviews = db.execute(
        select(FlowReview).where(FlowReview.record_id == record_id).order_by(FlowReview.review_id)
    ).scalars().all()
    pushes = db.execute(
        select(PushRecord).where(PushRecord.record_id == record_id).order_by(PushRecord.push_id)
    ).scalars().all()

    return {
        "flow": {
            "record_id": flow.record_id,
            "dedup_key": flow.dedup_key,
            "txn_no": flow.txn_no,
            "txn_date": flow.txn_date.isoformat() if isinstance(flow.txn_date, date) else str(flow.txn_date),
            "txn_time": flow.txn_time.isoformat(timespec="seconds") if isinstance(flow.txn_time, time) else None,
            "currency": flow.currency,
            "amount": str(flow.amount),
            "dc_flag": flow.dc_flag,
            "counterparty_name": flow.counterparty_name,
            "counterparty_account": flow.counterparty_account,
            "summary": flow.summary,
            "process_status": flow.process_status,
            "validation_status": flow.validation_status,
            "exception_type": flow.exception_type,
        },
        "batch": {
            "batch_id": batch.batch_id,
            "batch_no": batch.batch_no,
            "source_type": batch.source_type,
            "source_ref": batch.source_ref,
            "contract_version": batch.contract_version,
            "imported_by": batch.imported_by,
            "imported_at": batch.imported_at.isoformat() if isinstance(batch.imported_at, datetime) else str(batch.imported_at) if batch.imported_at else None,
        },
        "bank": {"bank_code": bank.bank_code, "bank_name": bank.bank_name} if bank else None,
        "account": {"account_no": account.account_no, "account_name": account.account_name} if account else None,
        "raw": {"raw_id": raw.raw_id, "raw_hash": raw.raw_hash, "raw_content": raw.raw_content} if raw else None,
        "validations": [
            {"rule_code": v.rule_code, "rule_result": v.rule_result, "error_detail": v.error_detail}
            for v in validations
        ],
        "reviews": [
            {"reviewer": r.reviewer, "review_result": r.review_result,
             "review_time": r.review_time.isoformat() if isinstance(r.review_time, datetime) else str(r.review_time),
             "matched_subject": r.matched_subject, "comment": r.comment}
            for r in reviews
        ],
        "pushes": [{"push_id": p.push_id, "push_status": p.push_status,
                    "voucher_no": p.voucher_no, "kingdee_doc_no": p.kingdee_doc_no,
                    "pushed_by": p.pushed_by, "error_msg": p.error_msg,
                    "pushed_at": p.pushed_at.isoformat() if isinstance(p.pushed_at, datetime) else str(p.pushed_at) if p.pushed_at else None} for p in pushes],
    }

File: backend/web/js/trace.js
// 共享：流水详情抽屉（契约字段 / 校验 / 溯源 / 留痕）+ 状态机操作按钮
import { api } from "./api.js";
import * as ui from "./ui.js";

const RULE_LABEL = {
  R001: "重复流水", R002: "负金额/方向非法", R003: "缺字段",
  R004: "超阈值", R005: "币种异常",
};

function field(k, v) {
  return `<div class="field"><span class="k">${ui.esc(k)}</span><span class="v">${v}</span></div>`;
}

function amtSpan(dc, amount) {
  const c = dc === "C" ? "in" : "out";
  const sgn = dc === "C" ? "+" : "−";
  return `<span class="amt ${c}">${sgn}¥ ${ui.money(amount)}</span>`;
}

function renderTrace(tr) {
  const f = tr.flow || {};
  const bank = tr.bank || {};
  const account = tr.account || {};
  const batch = tr.batch || {};

  const dir = f.dc_flag === "C" ? "收（贷方）" : f.dc_flag === "D" ? "付（借方）" : "—";
  const fieldsHtml = [
    field("唯一流水号", `<span class="mono">${ui.esc(f.txn_no || "—")}</span>`),
    field("交易日期", ui.fmtDate(f.txn_date)),
    field("交易时间", ui.esc(f.txn_time || "—")),
    field("借贷方向", dir),
    field("金额", amtSpan(f.dc_flag, f.amount)),
    field("币种", ui.esc(f.currency || "CNY")),
    field("对方户名", ui.esc(f.counterparty_name || "—")),
    field("对方账号", ui.esc(f.counterparty_account || "—")),
    field("本方账户", ui.esc(account.account_name ? `${account.account_name}（${account.account_no}）` : (account.account_no || "—"))),
    field("摘要", f.summary ? ui.esc(f.summary) : '<span style="color:var(--muted)">缺失</span>'),
  ].join("");

  const validations = tr.validations || [];
  let vHtml;
  if (validations.length === 0) {
    vHtml = `<div class="muted" style="font-size:13px">无校验记录</div>`;
  } else {
    vHtml = validations.map((v) => {
      const map = { PASS: ["通过", "var(--ok)"], WARN: ["告警", "var(--warn)"], FAIL: ["异常", "var(--danger)"] };
      const [label, color] = map[v.rule_result] || ["—", "var(--muted)"];
      const rule = RULE_LABEL[v.rule_code] ? `${v.rule_code} · ${RULE_LABEL[v.rule_code]}` : v.rule_code;
      return `<div class="trace-item">
        <span class="t">${ui.esc(rule)}</span>
        <span class="c"><span style="color:${color};font-weight:600">${label}</span>${v.error_detail ? `<div class="muted" style="margin-top:3px;font-size:12px">${ui.esc(v.error_detail)}</div>` : ""}</span>
      </div>`;
    }).join("");
  }

  const traceHtml = [
    ["来源银行", bank.bank_name ? `${ui.esc(bank.bank_name)}（${ui.esc(bank.bank_code)}）` : "—"],
    ["来源账户", account.account_no ? `${ui.esc(account.account_name || "")} ${ui.esc(account.account_no)}` : "—"],
    ["采集批次", `${ui.esc(batch.batch_no || "—")} <span class="tag-mini">${ui.esc(batch.source_type || "")}</span>`],
    ["记录 ID", `<span class="mono">${ui.esc(String(f.record_id ?? "—"))}</span>`],
    ["去重指纹", `<span class="mono" style="font-size:12px;word-break:break-all">${ui.esc(f.dedup_key || "—")}</span>`],
  ].map(([k, v]) => `<div class="trace-item"><span class="t">${k}</span><span class="c">${v}</span></div>`).join("");

  // 操作留痕时间线
  const events = [];
  events.push({ cls: "green", who: "系统", act: `采集落库 · 校验预筛`, tm: batch.imported_at ? ui.fmtDateTime(batch.imported_at) : "—", desc: batch.source_type ? `源：${batch.source_type} · 批次 ${batch.batch_no || "—"}` : "" });
  (tr.reviews || []).forEach((rv) => {
    const cls = rv.review_result === "PASS" ? "green" : rv.review_result === "ADJUST" ? "" : "reject";
    events.push({ cls, who: rv.reviewer || "复核员", act: `人工复核 · ${rv.review_result}`, tm: ui.fmtDateTime(rv.review_time), desc: [rv.matched_subject, rv.comment].filter(Boolean).join(" · ") });
  });
  (tr.pushes || []).forEach((p) => {
    events.push({ cls: p.push_status === "SUCCESS" ? "green" : "", who: p.pushed_by || "系统", act: `推送金蝶 · ${p.push_status}`, tm: ui.fmtDateTime(p.pushed_at), desc: p.voucher_no ? `凭证号 ${p.voucher_no}` : (p.error_msg || "") });
  });
  events.sort((a, b) => (a.tm || "").localeCompare(b.tm || ""));
  const audit = events.map((e) => `
    <div class="it ${e.cls}">
      <div class="who">${ui.esc(e.who)}</div>
      <div class="act">${ui.esc(e.act)}</div>
      <div class="tm">${ui.esc(e.tm)}${e.desc ? " · " + ui.esc(e.desc) : ""}</div>
    </div>`).join("");

  return { fieldsHtml, vHtml, traceHtml, audit, flow: f };
}

export function openRecordDrawer(record_id, { onChanged } = {}) {
  ui.openDrawer({ title: "流水详情", body: ui.loadingHtml("加载详情…") });
  api.traceByRecord(record_id).then((tr) => {
    const f = tr.flow || {};
    const { fieldsHtml, vHtml, traceHtml, audit } = renderTrace(tr);

    document.getElementById("dw-title").textContent = `流水详情 · #${record_id}`;
    document.getElementById("dw-badge").innerHTML = ui.statusBadge(f.process_status);
    document.getElementById("dw-body").innerHTML = `
      <div class="section-title">契约字段 · 统一流水契约</div>
      <div class="field-grid">${fieldsHtml}</div>

      <div class="section-title">校验结果</div>
      ${vHtml}

      <div class="section-title">数据溯源</div>
      <div class="trace-list">${traceHtml}</div>

      <div class="section-title">操作留痕 · 审计日志</div>
      <div class="audit">${audit || '<div class="muted" style="font-size:13px">无记录</div>'}</div>`;

    bindFooter(f, record_id, onChanged);
  }).catch((e) => {
    document.getElementById("dw-body").innerHTML = ui.emptyHtml("加载失败：" + e.message);
    document.getElementById("dw-foot").innerHTML = `<button class="btn btn-ghost" id="dw-close-btn">关闭</button>`;
    document.getElementById("dw-close-btn").addEventListener("click", ui.closeDrawer);
  });
}

function bindFooter(f, record_id, onChanged) {
  const s = f.process_status;
  const foot = document.getElementById("dw-foot");
  const closeBtn = `<button class="btn btn-ghost" data-act="close">取消</button>`;

  const doPass = (andPush) => async () => {
    try {
      if (s === "REVIEW_READY") await api.decide(record_id, "PASS");
      ui.toast("已通过复核", "success");
      if (andPush) await doPushOnce(record_id, onChanged);
      else { ui.closeDrawer(); onChanged && onChanged(); }
    } catch (e) { ui.toast(e.message, "error"); }
  };
  const doReject = async () => {
    const reason = await ui.promptModal({ title: "驳回流水", placeholder: "请填写驳回原因（必填，用于审计留痕）", okText: "确认驳回", danger: true });
    if (reason === null) return;
    if (!reason) { ui.toast("驳回原因不能为空", "error"); return; }
    try {
      await api.decide(record_id, "REJECT", null, reason);
      ui.toast("已驳回", "info");
      ui.closeDrawer();
      onChanged && onChanged();
    } catch (e) { ui.toast(e.message, "error"); }
  };
  const doPush = async () => {
    if (s === "REVIEW_READY") {
      try { await api.decide(record_id, "PASS"); } catch (e) { ui.toast(e.message, "error"); return; }
    }
    await doPushOnce(record_id, onChanged);
  };

  if (s === "REVIEW_READY") {
    foot.innerHTML = `
      <button class="btn btn-danger" data-act="reject">驳回</button>
      ${closeBtn}
      <button class="btn btn-ok" data-act="pass-push">通过并入账</button>
      <button class="btn btn-pri" data-act="push">推送金蝶</button>`;
  } else if (s === "REVIEW_PASSED") {
    foot.innerHTML = `${closeBtn}<button class="btn btn-pri" data-act="push">推送金蝶</button>`;
  } else {
    foot.innerHTML = closeBtn;
  }

  foot.querySelector('[data-act="reject"]')?.addEventListener("click", doReject);
  foot.querySelector('[data-act="pass-push"]')?.addEventListener("click", doPass(true));
  foot.querySelector('[data-act="push"]')?.addEventListener("click", doPush);
  foot.querySelector('[data-act="close"]')?.addEventListener("click", ui.closeDrawer);
}

async function doPushOnce(record_id, onChanged) {
  const p = await api.pushRecord(record_id);
  ui.toast(`推送成功，凭证号 ${p.voucher_no || "—"}`", "success");
  ui.closeDrawer();
  onChanged && onChanged();
}
